File: 20-valid-parentheses/valid-parentheses.py
from collections import deque
class Solution:
    def isValid(self, s: str) -> bool:
        stack=deque()
        pairs={')':'(',']':'[','}':'{'}
        opening=set(pairs.values())
        balance=True

        for i in s:
            if i in opening:
                stack.append(i)
            elif i in pairs:
                if not stack or stack.pop()!=pairs[i]:
                    balance= False
        if balance and len(stack)==0:
            return True                
        else:
            return False    

        

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
    HashMap<Character,Character> map=new HashMap<>();
    Stack<Character> st=new Stack<>();
    map.put(']','[');
    map.put(')','(');
    map.put('}','{');
    for(char ch:s.toCharArray()){
        if(map.containsKey(ch)){
            if(!st.isEmpty() && map.get(ch)==st.peek()){
                st.pop();
            }
            else{
                return false;
                }
        }
        else{
            st.push(ch);
        }        
    }
    return st.isEmpty();

        
    }
}